packages/runtime/container-runtime/src/pendingStateManager.ts

```diff
@@ -165,12 +209,82 @@ export class PendingStateManager {
         this.pendingStates.push(pendingFlush);
     }
 
+    public rebaseAt(seqNum: number) {
+        const rebasePromises: Promise<void>[] = [];
+        // rebase initial ops at sequence number
+        while (!this.initialStates.isEmpty()) {
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            const nextState = this.initialStates.peekFront()!;
+            if (nextState.type === "message") {
+                if (nextState.referenceSequenceNumber > seqNum) {
+                    break; // nothing left to do at this sequence number
+                } else if (nextState.referenceSequenceNumber > 0 && nextState.referenceSequenceNumber < seqNum) {
+                    // this method should be called at every sequence number in order, so this should never happen
+                    if (nextState.clientSequenceNumber === this.initialClientSeqNum) {
```

Comment:
I'm not sure why you want to throw different messages. The first message is correct either way and correctly reflects the reality. I do not think clientSequenceNumber is interesting here at all. I think what you are trying to say is that if this.initialStates has to be sorted by reference sequence number and by clientSequenceNumber at the same time. And that's true, but I'd assert that in constructor by just going through every message - it's so much cleaner in telling the intent.